### ISSUE TITLE
Quadrature rules not supported for high-order/curved meshes created with Gmsh

### ISSUE THREAD

**Issue by etiennespieser** [mfem/mfem#3759]
Hi all,

I am using a variant of `navier_3dfoc.cpp` from the `Navier` miniapp that I have modified to compute the flow around an airfoil. I created a high-order/curved grid to correctly represent my NACA geometry using scripting with Gmsh (python script used: [gmshRodAirfoil.py](https://github.com/etiennespieser/Propeller_Gmsh/blob/main/gmshRodAirfoil.py), and [gmshToolkit.py](https://github.com/etiennespieser/Propeller_Gmsh/blob/main/gmshToolkit.py)).

When trying to run simulations with my curved mesh, I come across the error: 
```
Verification failed: (Q1D <= MQ) is false:
 --> Quadrature rules with more than 8 1D points are not supported!
 ... in function: void mfem::internal::quadrature_interpolator::TensorDerivatives(int, int, const mfem::DofToQuad&, const mfem::Vector&, mfem::Vector&) [with mfem::QVectorLayout VL = mfem::QVectorLayout::byNODES]
 ... in file: fem/qinterp/grad_by_nodes.cpp:117
```

I tried different combinations of mesh order ("mo") and element order ("so") to run my simulation. The table below summarises the cases where my code works ("ok") and the cases where it fails ("X"), displaying exactly the same error message as the one pasted above.

<p align="center">
<img src="https://github.com/mfem/mfem/assets/16777709/addb3381-ab75-4877-b47f-c9465b3676b0" width=50% height=40%)>
</p>

Does someone have any idea what caused this error and how to fix it?

Étienne

**Comment by jandrej** [comment]
Hi, this is related to https://github.com/mfem/mfem/pull/3722. You can see the changes made for `MAX_D1D` and `MAX_Q1D` and adapt it yourself until we adopted a more permanent solution.

**Comment by v-dobrev** [comment]
Hi @etiennespieser,

My suggestion will be to instantiate the values you need (by adding them in the MFEM source) similar to here:  https://github.com/mfem/mfem/blob/2fafbbb2ef3bffd1c19b38b3b91ee01534d9aeae/fem/qinterp/grad_by_nodes.cpp#L114 The meaning of the dispatch parameter `id` in the `switch` statement can be seen in its definition here: https://github.com/mfem/mfem/blob/2fafbbb2ef3bffd1c19b38b3b91ee01534d9aeae/fem/qinterp/grad_by_nodes.cpp#L46

If you get similar errors in other places, you can do the same -- instantiate the optimized kernel versions for the parameters you need.

We realize that this way of modifying the MFEM source is not the most convenient way of solving this issue -- we are working on different approaches to address this, e.g. see PRs #3138 and #719.

**Comment by etiennespieser** [comment]
I instantiated the values in the source code up to `D1D=15` and `Q1D=15`, and modified the corresponding hard-coded values of `MD` and `QD` used in `MFEM_VERIFY()`. After re-building MFEM, I was able to run simulations for an order 6 curved mesh with order 6 elements. 

In my application relying on the Navier miniapp, starting from `mfem-4.5.2`, I had to modify `eval_by_nodes.cpp` (~ line 130), `grad_by_nodes.cpp` (~ line 110), and `nonlininteg_vectorconvection.cpp` (~ line 830). Also, I updated `forall.hpp` (~ line 30) to increase the values of `MAX_D1D` and `MAX_D1D`.

Many thanks to both of you for your help !

Étienne
